drivers/led_strip/ws2812_uart.c

```diff
@@ -0,0 +1,336 @@
+#define UART_FRAME_BITS_FROM_DT(inst) \
+	(1 + _DT_UART_DATA_BITS(inst) + _DT_UART_HAS_PARITY(inst) + _DT_UART_STOP_BITS(inst))
+
+/* Calculate the buffer size needed. */
+#define WS2812_UART_CALC_BUFSZ(num_px, num_colors, bits_symbol, bits_frame)                        \
+	DIV_ROUND_UP((num_px) * (num_colors) * BITS_PER_COLOR_CHANNEL * (bits_symbol),             \
+		     (bits_frame))
+
+struct ws2812_uart_cfg {
+	const struct device *uart_dev;
+	uint8_t *px_buf;
+	uint16_t one_symbol;
+	uint16_t zero_symbol;
+	uint8_t bits_per_symbol;
+	uint8_t num_colors;
+	const uint8_t *color_mapping;
+	size_t length;
+	uint16_t reset_delay;
+	uint8_t uart_frame_bits;
+};
+
+struct ws2812_uart_data {
+	struct k_mutex lock;
+	struct k_sem tx_done_sem;
+};
+
+/*
+ * Serializes an 8-bit color value into the UART buffer. This function takes
+ * an 8-bit color value, expands each of its 8 bits into the appropriate symbol
+ * pattern, and packs the resulting stream into UART data payloads.
+ */
+static inline void ws2812_uart_ser(uint8_t color, const struct ws2812_uart_cfg *cfg,
+				   uint8_t *frame_bit_pos, uint8_t **buf)
+{
+	for (int i = BITS_PER_COLOR_CHANNEL - 1; i >= 0; i--) {
+		uint16_t pattern = (color & BIT(i)) ? cfg->one_symbol : cfg->zero_symbol;
+
+		for (int p = cfg->bits_per_symbol - 1; p >= 0; p--) {
+			uint8_t pos = *frame_bit_pos;
+			/* Start and stop bits are always handled by hardware and skipped. */
+			bool is_hw_bit = (pos == 0) || (pos == (cfg->uart_frame_bits - 1));
+
+			/*
+			 * With an inverted signal, a high pulse ('1') is made by sending
+			 * a low level ('0'). We clear the bit as the buffer is pre-filled.
+			 */
+			if (!is_hw_bit && (pattern & BIT(p))) {
+				/* Map frame position to data position (no start bit). */
+				**buf &= ~BIT(pos - 1);
+			}
+
+			(*frame_bit_pos)++;
+			if (*frame_bit_pos >= cfg->uart_frame_bits) {
+				(*buf)++;
+				*frame_bit_pos = 0;
+			}
+		}
+	}
+}
+
+/*
+ * Callback for UART ASYNC API events.
+ */
+static void ws2812_uart_callback(const struct device *dev, struct uart_event *evt, void *user_data)
+{
+	struct k_sem *tx_done_sem = user_data;
+
+	if (evt->type == UART_TX_DONE) {
+		k_sem_give(tx_done_sem);
+	}
+}
+
+/*
+ * Latch current color values on strip and reset its state machines.
+ */
+static inline void ws2812_reset_delay(uint16_t delay)
+{
+	k_usleep(delay);
+}
+
+static int ws2812_strip_update_rgb(const struct device *dev, struct led_rgb *pixels,
+				   size_t num_pixels)
+{
+	const struct ws2812_uart_cfg *cfg = dev->config;
+	struct ws2812_uart_data *data = dev->data;
+	const size_t buf_len = WS2812_UART_CALC_BUFSZ(num_pixels, cfg->num_colors,
+						      cfg->bits_per_symbol, cfg->uart_frame_bits);
+	uint8_t *px_buf = cfg->px_buf;
+	uint8_t *current_buf = px_buf;
+	uint8_t frame_bit_pos = 0;
+	int ret;
+
+	/* Lock the driver to ensure thread-safe access to the buffer and UART */
+	k_mutex_lock(&data->lock, K_FOREVER);
```

Comment:
can this be a semaphore?

Reply:
Sempahores are normally preferred for driver code, few reasons for that, Andy goes in more details in https://youtu.be/p8OgqVQxklo?si=iaVjKvUwvntmViGn&t=1304 but yeah if you don't need priority inversion semaphore seems to be the way to go, also since most core code do not use mutexes the code has a chance of being compiled out if we keep not using it in drivers.